numpy/_core/src/multiarray/stringdtype/dtype.c

```diff
@@ -0,0 +1,910 @@
+/* The implementation of the StringDType class */
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include "structmember.h"
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+
+#include "numpy/arrayobject.h"
+#include "numpy/ndarraytypes.h"
+#include "numpy/npy_math.h"
+
+#include "static_string.h"
+#include "dtypemeta.h"
+#include "dtype.h"
+#include "casts.h"
+#include "gil_utils.h"
+#include "conversion_utils.h"
+#include "npy_import.h"
+
+/*
+ * Internal helper to create new instances
```

Comment:
This follows a pattern used by the dtypes I was cribbing off of in the `numpy-user-dtypes` repo. I probably could use `tp_init` for this but I'd rather not do that refactoring now since this function is used all over the place. Added it to the list of things I'll follow up on post-merge.